3_train/src/train.py

```diff
@@ -0,0 +1,465 @@
+import numpy as np
+import torch
+import torch.nn as nn
+from torch.utils.data import Dataset, DataLoader, random_split
+from torch.optim import Adam
+
+import sys
+sys.path.insert(1, '3_model/src')
+from models import Model
+
+
+class SequenceDataset(Dataset):
+    """
+    Custom Dataset class for sequences that include dynamic and static inputs,
+    and multiple outputs
+    
+    __len__ returns the number of sequences
+    
+    __getitem__ returns dynamic input features, static input features, and
+    temperature observations at all depths for one sequence
+    
+    :Example:
+
+
+    >>> npz = np.load("2_process/out/mntoha/train.npz")
+    >>> n_depths = len(npz['depths'])
+    >>> n_dynamic = len(npz['dynamic_features_all'])
+    >>> n_static = len(npz['static_features_all'])
+    >>> data = torch.from_numpy(npz['data'])
+    >>> dataset = SequenceDataset(data, n_depths, n_dynamic, n_static)
+    >>> # number of sequences
+    >>> len(dataset)
+    19069
+    >>> # dynamic features, static features, and temperatures at depths
+    >>> # for first sequence
+    >>> for e in dataset[0]:
+    >>>     print(e.shape)
+    >>>
+    torch.Size([400, 9])
+    torch.Size([400, 4])
+    torch.Size([400, 49])
+    """
+    def __init__(self, sequences, n_outputs, n_dynamic, n_static):
+        """
+        :param sequences: An array of sequences with shape (# sequences, sequence_length, # depths + # input features)
+        :param n_outputs: Number of outputs (# depths)
+        :param n_dynamic: Number of dynamic input features
+        :param n_static: Number of static input features
+
+        """
+        # The number of sequences is the size of the first dimension
+        self.len = len(sequences)
+        self.n_outputs = n_outputs
+        self.n_dynamic = n_dynamic
+        self.n_static = n_static
+        self.sequences = sequences
+
+    def __len__(self):
+        return self.len
+
+    def __getitem__(self, idx):
+        # sequence is organized as temperatures at multiple depths, then dynamic features, then static features
+        temperatures = self.sequences[idx, :, :self.n_outputs]
+        dynamic_features = self.sequences[idx, :, self.n_outputs:self.n_outputs + self.n_dynamic]
+        static_features = self.sequences[idx, :, -self.n_static:]
+        return dynamic_features, static_features, temperatures
+
+
+def split(dataset, fraction, seed):
+    """
+    Randomly split a dataset into two non-overlapping subsets
+
+    :param dataset: A torch.utils.data.Dataset to be split
+    :param fraction: The fraction of data samples to put in the first subset
+    :param seed: Integer seed for random number generator
+    :returns: Tuple of two datasets for the two subsets
+
+    """
+    # split into two subsets
+    n_total = len(dataset)
+    n_1 = int(round(n_total * fraction))
+    n_2 = n_total - n_1
+    subset_1, subset_2 = random_split(
+        dataset, [n_1, n_2], generator=torch.Generator().manual_seed(seed))
```

Comment:
I like this setting the seed as an argument, haven't seen that before (I usually just set it at the top of the script, but that is dependent on the same order and number of random operations)

Reply:
I set the seed on line 411 too, so this approach might be a bit redundant. The nice thing is that the train/validation split will be predictable, apart from any other changes to the workflow.

Reply:
I don't think it's redundant at all! As a general take-away for myself, I was trying to say that it's a good idea because it's not fragile to some new operation (you may insert later in a generic script) occurring before it and eating up its random seed value/iteration.

Ex of why your way is good

![SeedAsArg](https://user-images.githubusercontent.com/18230221/163602468-2595fafa-5ff1-46a8-b89a-860840e86594.PNG)

Reply:
Nicely demonstrated!